File: scripts/link-app.mjs
#!/usr/bin/env node
// Drive shopify app config link --reset via node-pty
import pty from "node-pty";
import fs from "node:fs";

const logPath = "/tmp/wwsl/pty.log";
const log = (msg) => { fs.appendFileSync(logPath, `[${new Date().toISOString()}] ${msg}\n`); };
fs.writeFileSync(logPath, "");

const shell = process.env.SHELL || "/bin/bash";
const child = pty.spawn("/Users/omarshahban/.nvm/versions/node/v22.19.0/bin/node", ["/Users/omarshahban/.nvm/versions/node/v22.19.0/bin/shopify", "app", "config", "link", "--reset"], {
  name: "xterm-256color",
  cols: 220,
  rows: 60,
  cwd: process.cwd(),
  env: process.env,
});

let buffer = "";
let state = "create";
let lastAction = 0;

function send(text, nextState) {
  const now = Date.now();
  if (now - lastAction < 1500) return false;
  lastAction = now;
  log(`>>> state=${state} -> send ${JSON.stringify(text)} next=${nextState}`);
  child.write(text);
  state = nextState;
  buffer = "";
  return true;
}

child.onData((data) => {
  buffer += data;
  const clean = data.replace(/\x1b\[[0-9;?]*[a-zA-Z]/g, "").replace(/\r/g, "");
  process.stdout.write(data);
  fs.appendFileSync(logPath, clean);

  const cleanBuf = buffer.replace(/\x1b\[[0-9;?]*[a-zA-Z]/g, "");

  setTimeout(() => {
    if (state === "create" && /Create this project as a new app/.test(cleanBuf)) {
      send("y", "name");
    } else if (state === "name" && /App name/.test(cleanBuf)) {
      send("Store Locator Maps WW Dev\r", "org");
    } else if (state === "org" && /Which (organization|developer organization|partner organization|account)/.test(cleanBuf)) {
      send("\r", "store");
    } else if (state === "store" && /Which (store|dev store|development store)/.test(cleanBuf)) {
      send("\r", "config");
    } else if (state === "config" && /Configuration file name|name for this configuration/.test(cleanBuf)) {
      send("\r", "done");
    }
  }, 500);
});

child.onExit(({ exitCode, signal }) => {
  log(`EXIT code=${exitCode} signal=${signal} state=${state}`);
  process.exit(exitCode || 0);
});

setTimeout(() => {
  log(`TIMEOUT state=${state}`);
  child.kill();
  process.exit(2);
}, 120000);

File: scripts/link-app.py
#!/usr/bin/env python3
"""Drive shopify app config link --reset via PTY. Send text and Enter as separate writes."""
import os, pty, select, re, sys, time

LOG = open("/tmp/wwsl/py.log", "w")
def log(m):
    LOG.write(f"[{time.time():.2f}] {m}\n"); LOG.flush()

SHOPIFY = "/Users/omarshahban/.nvm/versions/node/v22.19.0/bin/shopify"

pid, fd = pty.fork()
if pid == 0:
    os.environ["COLUMNS"] = "220"
    os.environ["LINES"] = "60"
    os.environ["TERM"] = "xterm-256color"
    os.execvp(SHOPIFY, [SHOPIFY, "app", "config", "link", "--reset"])

buffer = ""
state = "create"
last_send = 0
ANSI_RE = re.compile(r"\x1b\[[0-9;?]*[a-zA-Z]|\x1b[()][0-9A-Z]|\r")
deadline = time.time() + 240

def send_raw(text):
    os.write(fd, text.encode())

def send_text_then_enter(text):
    send_raw(text)
    time.sleep(0.6)
    send_raw("\r")

def advance(action_fn, next_state):
    global state, last_send, buffer
    now = time.time()
    if now - last_send < 2.0:
        return
    last_send = now
    log(f">>> state={state} -> {next_state}")
    action_fn()
    state = next_state
    buffer = ""

try:
    while time.time() < deadline:
        r, _, _ = select.select([fd], [], [], 0.5)
        if r:
            try:
                data = os.read(fd, 4096)
            except OSError:
                break
            if not data:
                break
            decoded = data.decode(errors="replace")
            sys.stdout.write(decoded); sys.stdout.flush()
            clean = ANSI_RE.sub("", decoded)
            LOG.write(clean); LOG.flush()
            buffer += clean
            if len(buffer) > 20000:
                buffer = buffer[-10000:]

        cb = buffer
        if state == "create" and "Create this project as a new app" in cb:
            advance(lambda: send_raw("y"), "name"); time.sleep(2); continue
        if state == "name" and "App name" in cb:
            advance(lambda: send_text_then_enter("StoreLocatorMapsWWDev"), "org"); time.sleep(2); continue
        if state == "org" and re.search(r"Which (organization|developer organization|partner organization|account)", cb):
            advance(lambda: send_raw("\r"), "store"); time.sleep(2); continue
        if state == "store" and re.search(r"Which (store|dev store|development store)", cb):
            advance(lambda: send_raw("\r"), "config"); time.sleep(2); continue
        if state == "config" and re.search(r"Configuration file name|name for this configuration", cb):
            advance(lambda: send_raw("\r"), "done"); time.sleep(2); continue

    log(f"LOOP END state={state}")
    LOG.close()
except Exception as e:
    log(f"ERROR {e}"); LOG.close(); raise
